service/history/decision/task_handler_test.go

```diff
@@ -0,0 +1,254 @@
+	"errors"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	"github.com/uber-go/tally"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/cache"
+	"github.com/uber/cadence/common/log/testlogger"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/persistence"
+	"github.com/uber/cadence/common/types"
+	"github.com/uber/cadence/common/types/testdata"
+	"github.com/uber/cadence/service/history/config"
+	"github.com/uber/cadence/service/history/constants"
+	"github.com/uber/cadence/service/history/execution"
+)
+
+const (
+	testTaskCompletedID = int64(123)
+)
+
+func TestHandleDecisionRequestCancelExternalWorkflow(t *testing.T) {
+	tests := []struct {
+		name            string
+		expectMockCalls func(taskHandler *taskHandlerImpl)
+		attributes      *types.RequestCancelExternalWorkflowExecutionDecisionAttributes
+		asserts         func(t *testing.T, taskHandler *taskHandlerImpl, err error)
+	}{
+		{
+			name: "success",
+			attributes: &types.RequestCancelExternalWorkflowExecutionDecisionAttributes{
+				Domain:            constants.TestDomainName,
+				WorkflowID:        constants.TestWorkflowID,
+				RunID:             constants.TestRunID,
+				Control:           nil,
+				ChildWorkflowOnly: false,
+			},
+			expectMockCalls: func(taskHandler *taskHandlerImpl) {
+				taskHandler.domainCache.(*cache.MockDomainCache).EXPECT().GetDomain(constants.TestDomainName).Return(constants.TestLocalDomainEntry, nil)
+			},
+			asserts: func(t *testing.T, taskHandler *taskHandlerImpl, err error) {
+				assert.False(t, taskHandler.failDecision)
+				assert.Empty(t, taskHandler.failMessage)
+				assert.Nil(t, taskHandler.failDecisionCause)
+				assert.Equal(t, nil, err)
+
+			},
+		},
+		{
+			name: "internal service error",
+			attributes: &types.RequestCancelExternalWorkflowExecutionDecisionAttributes{
+				Domain:            constants.TestDomainName,
+				WorkflowID:        constants.TestWorkflowID,
+				RunID:             constants.TestRunID,
+				Control:           nil,
+				ChildWorkflowOnly: false,
+			},
+			expectMockCalls: func(taskHandler *taskHandlerImpl) {
+				taskHandler.domainCache.(*cache.MockDomainCache).EXPECT().GetDomain(constants.TestDomainName).Return(nil, errors.New("some error getting domain cache"))
+			},
+			asserts: func(t *testing.T, taskHandler *taskHandlerImpl, err error) {
+				assert.Equal(t, &types.InternalServiceError{Message: "Unable to cancel workflow across domain: some random domain name."}, err)
+			},
+		},
+		{
+			name: "attributes validation failure",
+			asserts: func(t *testing.T, taskHandler *taskHandlerImpl, err error) {
+				assert.True(t, taskHandler.failDecision)
+				assert.NotNil(t, taskHandler.failMessage)
+				assert.Equal(t, func(i int32) *types.DecisionTaskFailedCause {
+					cause := new(types.DecisionTaskFailedCause)
+					*cause = types.DecisionTaskFailedCause(i)
+					return cause
+				}(9), taskHandler.failDecisionCause)
```

Comment:
This looks like a trick. Why two assignments are required? And where `9` is coming from?